### PR TITLE
Add fix for nested (Debug-Runspace) debugging on non-Windows platforms and re-enable Windows identity impersonation flow

### PR DESCRIPTION
**Pull request by PaulHigin**
## PR Summary

Nested runspace debugging is now supported on non-Windows platforms after Enter-PSHostProcess was added.  However, nested debugging was broken because the Windows impersonation code portion was not fully ifdefed out.

This change removes Windows only impersonation code for cross platform builds.  It also updates the Windows impersonation flow code to work with dotNet Core.

Repro:
```powershell
# On one PowerShell shell instance run:
$pid
14931

$ps = [powershell]::Create()
$ps.AddScript('1..20 | % { sleep 1; "Output $_" }').Invoke()

# On second PowerShell shell run:
Enter-PSHostProcess -Id 14931
[Process:14931]: PS > Debug-Runspace -Id 2

Expect:
Should be able to run a nested debug session on script.

Result:
Cannot debug script due to "Unsupported platform error".
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by TravisEz13 — APPROVED** [review]

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: How about organize the code this way:
```c#
            internal void DoInvoke()
            {
                try
                {
#if !UNIX
                    if (_identityToImpersonate != null)
                    {
                        WindowsIdentity.RunImpersonated(
                            _identityToImpersonate.AccessToken,
                            () => _results = _wrappedDebugger.ProcessCommand(_command, _output));
                        return;
                    }
#endif
                    _results = _wrappedDebugger.ProcessCommand(_command, _output);
                }
                catch (Exception e)
                {
                    _exception = e;
                }
                finally
                {
                    _commandCompleteEvent.Set();
                }
            }
        }
```

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: Another option is to only keep this if/def pair and remove the rest.
Referencing to `WindowsIdentity` is OK on Unix platform as long as we don't call methods on it (only then will the `PlatformNotSupportedException` be thrown).
This pair of if/def will make `_identityToImpersonate` always `null` on Unix, and thus we won't get in the code path that access any members of the `WindowsIdentity` type.

**Review by daxian-dbw — COMMENTED** [review]
@TravisEz13 @PaulHigin We ran into an issue before with `WindowsIdentity.RunImpersonated(safeAccessTokenHandle, action)` on Windows when porting powershell to CoreCLR, and it caused powershell to crash. We (us and .NET Core team) never figured out what was the root cause. I don't know if that issue is gone with the latest .NET Core. I will forward you the email thread on this.

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: Thanks, will do.

Also, I did some experiments and was able to force impersonation of a different account, and this (RunImpersonated) works without error.  I found four other places where we flow impersonation, so I'll make changes in this PR for those locations as well.

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: I am not sure what you mean, but I prefer to remove everything from UNIX builds that isn't used there.  I don't like having dead code paths...

**Comment by PaulHigin** [comment]
Looking through the code I realized there were many more places in the engine where Windows identity impersonation flow was disabled.  This is needed to support rare but important Windows scenarios.  I submitted a commit that re-enables these using the new dotNet RunImpersonated API.

I will also update the title to reflect these new changes.

**Comment by PaulHigin** [comment]
@daxian-dbw   All tests have passed.  Can this be merged?

File: src/System.Management.Automation/engine/Utils.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Utils.cs]: Our `CodeRefactor` rules require the `System` using directives to be put before other using directives. It will be great If you can update here (and in other touched files in this PR). If not, I will update it before merging.

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: This check doesn't seem to be needed. When using this method, we already know `_identityToImpersonate` is not null because `Utils.TryGetWindowsImpersonatedIdentity(out _identityToImpersonate)` returns true.

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Same here, this can be simplified to `() => InvokeThreadProc()`.

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Why not using `invokeThreadProcDelegate` here?
By looking at the code, calling `InvokeThreadProc` here would honor the `FlowImpersonationPolicy ` as well in Windows PowerShell.

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Same here. Why not use `invokeThreadProcDelegate` here?

File: src/System.Management.Automation/engine/remoting/fanin/WSManTransportManager.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/remoting/fanin/WSManTransportManager.cs]: > () => { result = WSManNativeApi.WSManDeinitialize(_handle, 0); }

This can be simplified to `() => result = WSManNativeApi.WSManDeinitialize(_handle, 0)`

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: How about using the other overload? 

```
_results = WindowsIdentity.RunImpersonated(_identityToImpersonate.AccessToken, () => _wrappedDebugger.ProcessCommand(_command, _output));
```

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Question: I see the `identityToImpersonate` is disposed at some places and not at some other places. For example, this instance is never disposed after use. Shall we always dispose it after use?

File: src/System.Management.Automation/engine/Utils.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/Utils.cs]: Oh ... didn't know that.  Will fix.

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Good point.  I forgot to remove this when refactoring...

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: Much better!

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Actually, I changed my mind about this.  Impersonation flow only needs to be performed when running on a *new* thread.  If the same thread is used, there is no need to run as impersonated regardless of whether the thread contains impersonation.  I feel the code expresses this by always using ThreadProc() when not started on a new thread.

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Yes, since WindowsIdentity has a Dispose() method I assume it has system resources to release.  So I call it whenever I feel I can do it safely to release system resources as soon as possible.  For other cases, like LocalPipeline, I am not sure when it is safe to dispose and I let finalizer thread do it ... I don't want to get an "invalid handle" exception.

For WSManTransportManger, I worry about multiple calls to the Dipose() method causing a handle exception, so I again don't dispose the WindowsIdentity object.  But this should be Ok since it is only used at the end of a remote session and finalizer should do the job.

File: src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs]: Good to know.

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Pull request merged**
